File: 107-Binary-Tree-Level-Order-Traversal-II/solution.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(root==null) return res;
        dfs(res,root,0);
        return res;
    }
    
    public void dfs(ArrayList<ArrayList<Integer>> res,TreeNode root, int k) {
        if(root==null) return;
        if(k==res.size()) res.add(0,new ArrayList<Integer>());
        res.get(res.size()-k-1).add(root.val);
        dfs(res,root.left,k+1);
        dfs(res,root.right,k+1);
    }
}

File: 107-Binary-Tree-Level-Order-Traversal-II/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = []
        self.helper(root, result, 0)
        return result[::-1]
    
    def helper(self, cur, result, depth):
        if not cur:
            return result
        if depth >= len(result):
            result.append([cur.val])
        else:
            result[depth].append(cur.val)
        self.helper(cur.left, result, depth + 1)
        self.helper(cur.right, result, depth + 1)